src/servers/src/otlp.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use api::v1::{InsertRequest, InsertRequests};
+use common_grpc::writer::{LinesWriter, Precision};
+use opentelemetry_proto::tonic::collector::metrics::v1::ExportMetricsServiceRequest;
+use opentelemetry_proto::tonic::common::v1::{any_value, KeyValue};
+use opentelemetry_proto::tonic::metrics::v1::{metric, number_data_point, *};
+use snafu::ResultExt;
+
+use crate::error::{self, Result};
+
+const GREPTIME_TIMESTAMP: &str = "greptime_timestamp";
+const GREPTIME_VALUE: &str = "greptime_value";
+
+/// Normalize otlp instrumentation, metric and attribute names
+///
+/// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-name-syntax
+/// - since the name are case-insensitive, we transform them to lowercase for
+/// better sql usability
+/// - replace `.` and `-` with `_`
+fn normalize_otlp_name(name: &str) -> String {
+    name.to_lowercase().replace(|c| c == '.' || c == '-', "_")
+}
+
+/// Convert OpenTelemetry metrics to GreptimeDB insert requests
+///
+/// See
+/// https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L162
+/// for data structure of OTLP metrics.
+///
+/// Returns `InsertRequests` and total number of rows to ingest
+pub fn to_grpc_insert_requests(
+    request: ExportMetricsServiceRequest,
+) -> Result<(InsertRequests, usize)> {
+    let metrics = request
+        .resource_metrics
+        .iter()
+        .flat_map(|resource_metrics| &resource_metrics.scope_metrics)
+        .flat_map(|scope_metrics| &scope_metrics.metrics);
+
+    let mut insert_batch = Vec::with_capacity(metrics.size_hint().0);
+    for metric in metrics {
+        if let Some(insert) = encode_metrics(metric)? {
+            insert_batch.push(insert);
+        }
+    }
+
+    let rows = insert_batch.iter().map(|i| i.row_count as usize).sum();
+    let inserts = InsertRequests {
+        inserts: insert_batch,
+    };
+
+    Ok((inserts, rows))
+}
+
+fn encode_metrics(metric: &Metric) -> Result<Option<InsertRequest>> {
```

Comment:
Did I get it right that `Result<Option<InsertRequest>>` return type will be handy for the next version of the function so it will allow more fine-grained error handling?

Reply:
Normally it should be `Result<InsertRequest>` here but according to the definition of OTLP's protobuf, there is a chance that the metric carries no `data` so I have to add an `Option` type here.

Reply:
Thanks for the explanation 🙏